### PR TITLE
[DLS] Access control sync scheduling

### PR DESCRIPTION
**Pull request by timgrein**
## Closes https://github.com/elastic/enterprise-search-team/issues/4690

This PR adds scheduling for access control sync jobs. Access control sync jobs can only be scheduled, if the corresponding DLS feature flag is enabled. It uses the same concurrency control mechanisms as other job types.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: connectors/protocol/connectors.py [hunk 1]
```diff
@@ -869,7 +878,7 @@ def _create_object(self, doc_source):
             doc_source=doc_source,
         )
 
-    async def create(self, connector, trigger_method):
+    async def create(self, connector, trigger_method, job_type):
```
Comment (wangch079) [inline connectors/protocol/connectors.py:881]: This was done in PR #995

File: connectors/protocol/connectors.py [hunk 2]
```diff
@@ -511,7 +512,7 @@ def incremental_sync_scheduling(self):
         return self.get("scheduling", "incremental", default={})
 
     @property
-    def access_control_scheduling(self):
+    def access_control_sync_scheduling(self):
```
Comment (wangch079) [inline connectors/protocol/connectors.py:515]: This can be a separate PR. But I'm OK to review it in this PR.

File: connectors/services/job_scheduling.py [hunk 3]
```diff
@@ -28,6 +31,44 @@
 from connectors.source import get_source_klass
 
 
+class LastSyncTimeMethods(Enum):
```
Comment (wangch079) [inline connectors/services/job_scheduling.py:34]: Instead of introducing a new class. WDYT if we introduce 2 new methods in `Connector` class?
```python
def last_sync_scheduled_at_by_job_type(job_type):
    match job_type:
        case JobType.ACCESS_CONTROL:
            return self.last_access_control_sync_scheduled_at
        case JobType.INCREMENTAL:
            return self.last_incremental_sync_scheduled_at
        case JobType.FULL:
            return self.last_full_sync_scheduled_at
        case _:
            raise ValueError(f"Unknown job type: {job_type}")

async def update_last_sync_scheduled_at_by_job_type(job_type, new_ts):
    match job_type:
        case JobType.ACCESS_CONTROL:
            await self._update_datetime("last_access_control_sync_scheduled_at", new_ts)
        case JobType.INCREMENTAL:
            await self._update_datetime("last_incremental_sync_scheduled_at", new_ts)
        case JobType.FULL:
            await self._update_datetime("last_sync_scheduled_at", new_ts)
        case _:
            raise ValueError(f"Unknown job type: {job_type}")
```

**Review by wangch079 — COMMENTED** [review]
For access control sync, do you need to check the feature flag? If it's not enabled, we should skip directly.

File: connectors/services/job_scheduling.py (hunk 3 above)
Comment (timgrein) [reply connectors/services/job_scheduling.py:34]: Makes sense! Increases the cohesion and removes the "magic" lookups. Will adjust it 👍

File: connectors/protocol/connectors.py (hunk 1 above)
Comment (timgrein) [reply connectors/protocol/connectors.py:881]: Will fetch your changes as soon as they're on main 👍

File: tests/protocol/test_connectors.py [hunk 4]
```diff
@@ -543,24 +543,44 @@ async def test_sync_done(job, expected_doc_source_update):
...
 @patch("connectors.protocol.connectors.next_run")
-async def test_connector_next_sync(next_run, scheduling_enabled, expected_next_sync):
+async def test_connector_next_sync(next_run, scheduling_enabled, expected_next_sync, job_type):
     connector_doc = {
         "_id": "1",
         "_source": {
-            "scheduling": {"full": {"enabled": scheduling_enabled, "interval": "1 * * * * *"}},
+            "scheduling": {
+                "enabled": scheduling_enabled,
+                "interval": "1 * * * * *",
```
Comment (wangch079) [inline tests/protocol/test_connectors.py:563]: related to this comment: https://github.com/elastic/enterprise-search-team/issues/4589#issuecomment-1567909683

Why we want to keep the outer `enabled` and `interval`?

**Comment by wangch079** [comment]
Hi @timgrein , I think you missed this comment: https://github.com/elastic/connectors-python/pull/1000#pullrequestreview-1456737985

File: tests/protocol/test_connectors.py (hunk 4 above)
Comment (timgrein) [reply tests/protocol/test_connectors.py:563]: We don't need it here, but it'll be kept in the mapping to avoid reindex operations on a migration. See the corresponding discussion inside the implementation doc, but I can remove it from this test 👍

**Comment by timgrein** [comment]
> Hi @timgrein , I think you missed this comment: [#1000 (review)](https://github.com/elastic/connectors-python/pull/1000#pullrequestreview-1456737985)

I've merged main into this branch, so the changes are present, but not in the diff (see [job_type param in create method](https://github.com/elastic/connectors-python/pull/1000/files#diff-c2f73fe85d1652beb4af8939b73360a399a3496c1863e34b900e12d694420571R898))

**Comment by timgrein** [comment]
> Hi @timgrein , I think you missed this comment: [#1000 (review)](https://github.com/elastic/connectors-python/pull/1000#pullrequestreview-1456737985)

Wanted to address it in a separate PR originally, but I've added it now to this PR as it's related 👍

File: tests/protocol/test_connectors.py (hunk 4 above)
Comment (wangch079) [reply tests/protocol/test_connectors.py]: I think they are removed from both the [mapping](https://github.com/elastic/ent-search/blob/63ce9682bc92b9d57712e7e267550042f4cabca8/shared_togo/lib/shared_togo/crawler2/connector.rb#L156-L177) and the [migration](https://github.com/elastic/ent-search/blob/63ce9682bc92b9d57712e7e267550042f4cabca8/shared_togo/db/migrate/20230526133312_change_scheduling_and_add_job_type_and_sync_cursor_to_connectors.rb#L61-L69)

File: connectors/services/job_scheduling.py [hunk 5]
```diff
@@ -171,48 +173,67 @@ async def _should_schedule_on_demand_sync():
...
                 await connector.reload()
             except DocumentNotFoundError:
                 logger.error(f"Couldn't reload connector {connector.id}")
                 return False
 
-            now = datetime.utcnow()
             if (
-                connector.last_sync_scheduled_at is not None
-                and connector.last_sync_scheduled_at > now
+                job_type == JobType.ACCESS_CONTROL
```
Comment (wangch079) [inline connectors/services/job_scheduling.py:184]: For incremental sync, I will check both the feature flag and this value:
https://github.com/elastic/connectors-python/blob/e4ac7dc0f6b4b674eefc1787868524c48a8085af/connectors/source.py#L342

Not sure whether this applies to DLS.

What do you think we move this check to line 99:
https://github.com/elastic/connectors-python/blob/e4ac7dc0f6b4b674eefc1787868524c48a8085af/connectors/services/job_scheduling.py#L94-L101

This way, we can 1) access `source_klass. support_incremental_sync` directly, 2) avoid `if...else...` in `_should_schedule_scheduled_sync`.

Comment (timgrein) [reply connectors/services/job_scheduling.py]: Makes sense, adjusted it and did some other simplifications 👍 

> Not sure whether this applies to DLS.

No, this doesn't apply to DLS. We can still discuss it, but this should be a separate change from this PR.

File: connectors/services/job_scheduling.py [hunk 6]
```diff
@@ -171,7 +173,7 @@ async def _should_schedule_on_demand_sync():
                 job_type=JobType.FULL,
             )
 
-    async def _scheduled_sync(self, connector):
+    async def _scheduled_sync(self, connector, job_type=None):
```
Comment (wangch079) [inline connectors/services/job_scheduling.py:176]: I think we can make `job_type` mandatory.

**Review by wangch079 — APPROVED** [review]
Nice work
